openedx/frontend-app-course-authoring
# Convert "Pages & Resources" page to a plugin system

This is a proof-of-concept with a real-world use case for the [Frontend Pluggability Summit](https://discuss.openedx.org/t/frontend-pluggability-summit/11167).

The "Pages & Resources" page of the Course Authoring MFE displays the settings for any "Course Apps" enabled for a particular course. You can read [in this ADR](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/course_apps/docs/decisions/001-course-apps.rst) more about Course Apps : 

> The new [Course Authoring MFE](https://github.com/openedx/frontend-app-course-authoring/) includes a new UX called "Pages and Resources" for configuring different aspects of the course experience such as progress, wiki, teams, discussions, etc. all from one place. ... This ADR describes how we can use the existing plugin architecture of Open edX to automatically discover such apps and expose them via an API. This API will be used to list the installed apps that are available for a course and to enable/disable these apps using the API.

However, even though Course Apps were specifically designed for this Course Authoring MFE and even though they use a plugin framework on the backend, the frontend is currently hard-coded, and every configurable course app must be built in to frontend-app-course-authoring. Including, for example, 2U's (proprietary?) ["Xpert Unit Summaries" configuration UI](https://github.com/openedx/frontend-app-course-authoring/tree/master/src/pages-and-resources/xpert-unit-summary).

**What this PR does**: is makes every "Course App" _frontend_ an installable plugin as well. Eight plugins are included in the repo and installed by default (`calculator`, `edxnotes`, `live`, `ora_settings`, `proctoring`, `progress`, `teams`, `wiki`), one is still built-in (`discussions` - because it has a totally different UI than the others), and one is still included in the repo for now but not installed by default (`xpert_unit_summary`).

![Screenshot 2023-10-17 at 11 48 33 AM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/39d6a5e1-f56e-4ffb-a361-986d853d0ff1)

A basic plugin with a fairly complex UI:

![Screenshot 2023-10-17 at 11 55 58 AM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/16ee290e-ddb8-48fb-a676-b9158d8b8c09)

A plugin showing custom styles (the "reset all units" button has custom SCSS that is part of the plugin):

![Screenshot 2023-10-17 at 11 54 27 AM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/92f5331a-9fd4-4ce9-b9c7-4d564cddf5cf)


### Technical Details ###

Plugins will be detected when they are `npm install`ed. For example, to enable the "xpert unit summary" plugin, you have to run the command `npm install some-other-repo/plugins/xpert_unit_summary/ --no-save`. I put that plugin in a folder called `some-other-repo` for now to show that it doesn't matter where the plugin is hosted - it can be a totally separate repo/npm-package, and yet it will still install and register with the MFE.

Because we have the course apps functionality on the backend, this particular usage of plugins does not have any need to "list all installed plugins" so you won't find any code for that. Instead, the backend determines which course apps are enabled, and then the MFE just assumes that the corresponding plugin is available and will load it on demand when the user clicks on the app card.

If you _did_ want to be able to list all the installed frontend plugins from the MFE code, it's actually quite easy to do though:

```javascript
/**
 * Load all installed plugins that match a given regex by scanning the
 * node_modules folder.
 *
 * Note that this is actually done by webpack at build time, so there's almost
 * no runtime overhead to calling this function.
 * 
 * @param {{(): unknown, keys: () => string[]}} loader result of calling require.context()
 * @param {RegExp} regex regular expression that will match the plugins installed in the @openedx-plugins npm namespace
 *                       must contain a capturing group around the plugin ID, like /course-app-(\w+)\/Settings\.jsx$/
 * @returns {{[pluginName: string]: () => Record<string, unknown>}}
 */
function loadPluginContext(loader, regex) {
    // Load a require context. Allows us to get a list of matching plugins without knowing them in advance.
    // See https://webpack.js.org/guides/dependency-management/#requirecontext
    return Object.fromEntries(
        // convert from paths like "./course-app-calculator/Settings.jsx" to plugin names like "calculator"
        loader.keys().map(key => [regex.exec(key)[1], () => loader(key)])
    );
}

export const courseAppSettingsPlugins = loadPluginContext(
    // Due to how webpack works, the following strings need to be static, and we have to duplicate the regex.
    require.context('../node_modules/@openedx-plugins/', true, /course-app-(\w+)\/Settings\.jsx$/),
    /course-app-(\w+)\/Settings\.jsx$/,
);

// Access as e.g. courseAppSettingsPlugins["calculator"]()
// or list all installed plugins with Object.keys(courseAppSettingsPlugins)
```

### How to test it out

If you have the MFE installed and running in development mode already, just check out this branch. You'll need to run `npm install` within the MFE's root folder then restart the MFE. If you want to test the Xpert Unit Summaries, you'll need to have [the corresponding backend plugin](https://github.com/edx/ai-aside) installed, and then run `npm install plugins/course-apps/xpert_unit_summary/ --no-save` to install the frontend plugin into the MFE environment. To make things easier, I used [this shim](https://github.com/openedx/edx-platform/compare/master...open-craft:edx-platform:braden/force-enable-course-app-plugins) to fake the AI Aside/Xpert plugin and force enable many others.

### How would I use this in production?

You'd just have to add some additional step to `npm install @openedx-plugins/my-custom-plugin` for any custom plugins you have defined when you build the MFE.
